### DISCUSSION TITLE
base64 이미지가 전체 data에 적용되는 이슈 해결 / promise 와 forEach

### DISCUSSION THREAD

**Discussion by kjungit** [KDT4-team6/Paldo#24] [Talking]
전체 데이터를 넣는 코드에서 data.json파일에서 들고온 배열을 forEach로 순회했었는데요!
그 안에서 사용하던 base64이미지를 변환하는 함수내에서 promise 해결을 완료하기 전에 forEach가 순회를 진행하고 있기 때문에 마지막에 들어가는 데이터의 이미지가 전부 할당되는거 였습니다.

그래서 for...of 문을 통해서 다음 loop로 넘어가기전 base64이미지를 변환하는하는 promise를 모두 기다린 후 각각 적용되었습니다.

만약 배열이나 여러 데이터를 순회하는 코드가 있다면 forEach보다는 for...of를 사용하는게 좋을 것 같습니다.

**Comment by yousunzoo** [comment]
forEach와 for ... of의 정확한 차이가 무엇일까요...?

**Reply by kjungit** [reply]
배열을 순회하는것은 같지만 promise를 만나게되면 다르게 실행되는것 같습니다.
forEach가 순회할때에는 promise가 await하기 전에 다음 순회로 넘어가게 됩니다!
for...of 는 다음 item으로 넘어가기전 await가 모두 끝날때까지 대기하고 끝난 후 넘어갑니다!

**Reply by kjungit** [reply]
https://velog.io/@hanameee/%EB%B0%B0%EC%97%B4%EC%97%90-%EB%B9%84%EB%8F%99%EA%B8%B0-%EC%9E%91%EC%97%85%EC%9D%84-%EC%8B%A4%EC%8B%9C%ED%95%A0-%EB%95%8C-%EC%95%8C%EC%95%84%EB%91%90%EB%A9%B4-%EC%A2%8B%EC%9D%84%EB%B2%95%ED%95%9C-%EC%9D%B4%EC%95%BC%EA%B8%B0%EB%93%A4

이 블로그글을 보게 되면 이해가 조금 더 쉬울것 같습니다!
forEach 동작원리에서 비동기 작업을 하는것을 신경쓰지 않기 때문에 비동기작업에서는 forEach를 지양하고, 순차처리가 필요한지 등을 고려해봐야할 것 같습니다!

**Reply by FranzCho** [reply]
공부할 것 추가,,, 메모,,,
